### PR TITLE
[bookmarks] Add option to process "Saved Places" GeoJSON, and export as GPX to Google Maps script

### PR DESCRIPTION
**Pull request by rudokemper**
Closes https://github.com/organicmaps/organicmaps/issues/9038.

I expanded the `google_maps_bookmarks.py` Python script to work with GeoJSON (e.g. what you get from retrieving a Saved Places.json file by navigating to "Your data in Maps"), and also to export as GPX just to have a non-proprietary output option.

I considered making this a separate script, but it's probably more confusing to have two Google Maps scripts than one which handles all of the Google Maps saved places exports.

**Details**

* The script now takes arguments `--input` to point to the file, and `--format` to specify kml or gpx (default).
* The `convert` method detects the mime type of the file to determine if it's a GeoJSON or CSV input, and makes other method calls accordingly.
* For Google Maps CSV files, we require an API key; it is not needed for the GeoJSON input of starred places.
* Added a `process_geojson_features` method, and a util function to render the timestamp more human legible. Note: here, I am largely adopting the same formatting of placing all of the Google maps metadata in `description` with bolded field names and line breaks, similar to how I handled that in JavaScript [here](https://github.com/rudokemper/google-maps-places-to-organic-maps/blob/e9a5d7e3e0074ed23e62472666bc7cf0d7402398/script/main.js#L6). Open to other suggestions.
* Added a `write_gpx` method to support that output format.

### CODE REVIEW

File: tools/python/google_maps_bookmarks.py [hunk 1]
```diff
@@ -2,97 +2,133 @@
...
 import urllib.parse
 import urllib.request
 import xml.etree.ElementTree as ET
-from os import R_OK
+from os import path, access, R_OK, linesep
+from io import StringIO
 
 
-class GoogleMapsConverter:
+from dateutil import parser as dateutil_parser
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:16]: It looks like `dateutil` is not part of the python distribution and the user has to install this module via `pip`. The script should run without any external dependencies. Could you replace `dateutil` with `datetime`?

File: tools/python/google_maps_bookmarks.py [hunk 2]
```diff
@@ -2,97 +2,133 @@
...
-        print()
-        print("1. Create an API key for Google Places API following this guide")
+class GoogleMapsConverter:
+    def __init__(self, input_file, output_format):
+        print("Follow these steps to export your saved places from Google Maps and convert them to a GPX or KML File")
         print()
-        print("https://developers.google.com/maps/documentation/places/web-service/get-api-key")
+        print("OPTION 1: Using GeoJSON")
+        print("===================================")
+        print("1. Go to \"Your data in Maps\" in your Google Account settings or by accessing https://myaccount.google.com/yourdata/maps")
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:24]: I think these instructions could confuse the users of this script. They just want to export their bookmarks from Google Maps, so their probably not thinking in a GeoJSON/CSV workflow. If you follow the initial instructions, you also get the GeoJSON file from Google Takeout. You also need an API-Key for GeoJSON (see https://github.com/organicmaps/organicmaps/pull/8721#pullrequestreview-2247250563). So I would keep the initial instructions and add a step 4a./4b. that explains how the script is called with a csv or geojson file.

File: tools/python/google_maps_bookmarks.py [hunk 3]
```diff
@@ -104,26 +140,51 @@ def write_kml(self):
...
+                raise ValueError(f"Unsupported file format: {self.input_file}")
+        
+        # Write to output file in the desired format, KML or GPX
+        if self.output_format == 'kml':
+            self.write_kml()
+        elif self.output_format == 'gpx':
+            self.write_gpx()
 
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Convert Google Maps saved places to KML or GPX.")
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:184]: The script should either be called with arguments or interactively. You could add `input` and `format` to the interactive input or set the bookmarks name and API key with arguments.

File: tools/python/google_maps_bookmarks.py [hunk 4]
```diff
@@ -2,97 +2,133 @@
...
-            if not bookmark_list_name:
-                print("Please provide a name" + os.linesep)
-                continue
-            else:
-                self.kml_file = bookmark_list_name + ".kml"
-                break
-        print()
-        self.places = []
+            properties = feature['properties']
+            location = properties.get('location', {})
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:86]: Sometimes the coordinates are set to zero in GeoJSON. See https://github.com/organicmaps/organicmaps/pull/8721#pullrequestreview-2247250563 and the comment below. How do you handle this case?

File: tools/python/google_maps_bookmarks.py [hunk 5]
```diff
@@ -104,26 +140,51 @@ def write_kml(self):
...
+        tree.write(self.output_file)
+        print("Exported Google Saved Places to " + path.abspath(self.output_file))
+
+    def convert(self):
+        with open(self.input_file, 'r') as file:
+            content = file.read().strip()
+            if not content:
+                raise ValueError(f"The file {self.input_file} is empty or not a valid JSON file.")
+            
+            # Determine the file mime type (GeoJSON or CSV) and process accordingly
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:163]: This comment is redundant. You can remove it.

File: tools/python/google_maps_bookmarks.py [hunk 6]
```diff
@@ -104,26 +140,51 @@ def write_kml(self):
...
+
+    def write_gpx(self):
+        gpx = ET.Element("gpx", version="1.1", creator="GoogleMapsConverter")
+        for place in self.places:
+            wpt = ET.SubElement(gpx, "wpt", lat=place['coordinates'].split(',')[1], lon=place['coordinates'].split(',')[0])
+            ET.SubElement(wpt, "name").text = place['name']
+            ET.SubElement(wpt, "desc").text = place['description']
+        tree = ET.ElementTree(gpx)
+        tree.write(self.output_file)
+        print("Exported Google Saved Places to " + path.abspath(self.output_file))
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:155]: This line is duplicated in both `write_*` methods. You could move it at the end of the `convert` method.

**Review by fabwu — CHANGES_REQUESTED** [review]
Thanks for adding support for GeoJSON. I added some comments but overall it looks fine.

File: tools/python/google_maps_bookmarks.py (hunk 4 above)
Comment (rudokemper) [reply tools/python/google_maps_bookmarks.py]: Thanks. As you correctly pointed out, for these cases we can request coordinates from the Google API. I've adapted the code to handle this for when we have a google maps URL with query param `q` (and no coords) or `cid`. I have tested this with my own GeoJSON and it works for all cases.

Comment (rudokemper) [reply tools/python/google_maps_bookmarks.py]: Correction: `cid` does not work in the way I had it, and as of yet I haven't figured out a way to convert those. So I've modified the script to skip those and print a note that the script doesn't yet support `cid`. Not ideal but maybe we can handle that in a follow-on PR.

File: tools/python/google_maps_bookmarks.py [hunk 7]
```diff
@@ -104,26 +243,55 @@ def write_kml(self):
...
+            
+            mime_type, _ = mimetypes.guess_type(self.input_file)
+            if mime_type == 'application/geo+json' or mime_type == 'application/json':
+                try:
+                    geojson = json.loads(content)
+                except json.JSONDecodeError:
+                    raise ValueError(f"The file {self.input_file} is not a valid JSON file.")
+                self.process_geojson_features(geojson)
+            elif mime_type == 'text/csv':
+                self.get_api_key()
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:271]: We call this function already in init so you can remove this call

File: tools/python/google_maps_bookmarks.py [hunk 8]
```diff
@@ -2,97 +2,236 @@
...
         while True:
+            self.bookmark_list_name = input("Bookmark list name: ")
+            if not self.bookmark_list_name:
+                print("Please provide a name" + linesep)
+                continue
+            else:
+                self.output_file = self.bookmark_list_name + "." + self.output_format
+                break
+            
+    def get_api_key(self):
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:86]: This method is defined twice

File: tools/python/google_maps_bookmarks.py [hunk 9]
```diff
@@ -104,26 +243,55 @@ def write_kml(self):
...
+    def convert(self):
+        with open(self.input_file, 'r') as file:
+            content = file.read().strip()
+            if not content:
+                raise ValueError(f"The file {self.input_file} is empty or not a valid JSON file.")
+            
+            mime_type, _ = mimetypes.guess_type(self.input_file)
+            if mime_type == 'application/geo+json' or mime_type == 'application/json':
+                try:
+                    geojson = json.loads(content)
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:266]: Could you move the json parsing into the `process_geojson_features`? Then it's more in line with the `process_csv_features` function, where the parsing is also done inside the method.

File: tools/python/google_maps_bookmarks.py [hunk 10]
```diff
@@ -2,97 +2,236 @@
...
+                return q.split(',')
+            else:
+                url = f"https://maps.googleapis.com/maps/api/place/textsearch/json?query={urllib.parse.quote(q)}&key={api_key}"
+                return None
+        elif cid:
+            print(f"This script does not yet support extracting coordinates from a CID. Skipping {cid}" )
+        else:
+            return None
+
+        if url:
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:139]: You can call the `get_json` function to get a json from your url. No need to write the code for fetching a json again.

File: tools/python/google_maps_bookmarks.py (hunk 4 above)
Comment (fabwu) [reply tools/python/google_maps_bookmarks.py]: The example here https://github.com/organicmaps/organicmaps/pull/8721#issuecomment-2298220602 has a `ftid`. Does your data has a `cid`?

Anyway, the csv code already converts a `ftid` to coordinates by calling `https://maps.googleapis.com/maps/api/place/details/json?ftid=...`. Maybe the same API works with the `cid` parameter?

File: tools/python/google_maps_bookmarks.py [hunk 11]
```diff
@@ -2,97 +2,236 @@
...
-                            print(f"Couldn't extract coordinates from Googe Maps. Skipping {name}")
-                            continue
+    def get_coordinates_from_google_api(self, api_key, q=None, cid=None):
+        url = None
+        if q:
+            # Sometimes the 'q' parameter is a comma-separated lat long pair
+            if ',' in q and all(part.replace('.', '', 1).replace('-', '', 1).isdigit() for part in q.split(',')):
+                return q.split(',')
+            else:
+                url = f"https://maps.googleapis.com/maps/api/place/textsearch/json?query={urllib.parse.quote(q)}&key={api_key}"
```
Comment (fabwu) [inline tools/python/google_maps_bookmarks.py:132]: You could use `urllib.parse.urlencode` for the params so you don't have to quote and join them manually

**Review by fabwu — CHANGES_REQUESTED** [review]

**Comment by rudokemper** [comment]
Thanks @fabwu for your code review and the suggestion to use this API route `https://maps.googleapis.com/maps/api/place/details/json` for `cid`: it worked!

The script is now able to process the majority of my own ~3,800 Saved Places.

**There are few places still for which the Google API does not return the coordinates**; some with a `q` param, and some with a `cid` param. Here is a GeoJSON that includes one of each:

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "geometry": {
        "coordinates": [
          0,
          0
        ],
        "type": "Point"
      },
      "properties": {
        "date": "2019-01-09T04:43:31Z",
        "google_maps_url": "http://maps.google.com/?q=Unnamed+Road,+2+Chome+Ueno,+Tait%C5%8D-ku,+T%C5%8Dky%C5%8D-to+110-0005,+Japan&ftid=0x60188c20f3688631:0xfa27732134def7d4",
        "Comment": "No location information is available for this saved place"
      },
      "type": "Feature"
    },
    {
      "geometry": {
        "coordinates": [
          0,
          0
        ],
        "type": "Point"
      },
      "properties": {
        "date": "2018-08-01T18:53:27Z",
        "google_maps_url": "http://maps.google.com/?cid=15807036382859735710",
        "Comment": "No location information is available for this saved place"
      },
      "type": "Feature"
    }
  ]
}
```

In the case of the first feature (with `q` param), the API returns ZERO_RESULTS. Since there is an `ftid` param provided, I tried to fetch that using the `https://maps.googleapis.com/maps/api/place/details/json` endpoint, which returned "NOT_FOUND: The provided Place ID is no longer valid. Please refresh cached Place IDs as per https://developers.google.com/places/web-service/place-id#save-id"

In the case of the second (with `cid` param), the API also returns the same message to refresh cached Place IDs. So we know that the issue with both features is an outdated Place ID.

I tried to see if we could solve that programmatically by following the instructions on the linked URL, which suggests refreshing stored place IDs using the Places API. Unfortunately, this did not prove fruitful:

* Places API (New): `https://places.googleapis.com/v1/places/15807036382859735710?fields=id&key=` returns "Not a valid Place ID: 15807036382859735710\"`
* Places API: `https://maps.googleapis.com/maps/api/place/details/json? place_id=15807036382859735710?fields=place_id&key=` returns "Invalid placeid parameter"

(Same result using the `ftid` from the first feature.)

I suspect that these instructions to refresh cached Place IDs don't apply to `ftid` or `cid`, so I'm not sure then how to access a valid Place ID for these.

Unless anyone has a suggestion, maybe we can leave this as an issue for follow-on work?

**Comment by fabwu** [comment]
> Thanks @fabwu for your code review and the suggestion to use this API route `https://maps.googleapis.com/maps/api/place/details/json` for `cid`: it worked!
> 
> The script is now able to process the majority of my own ~3,800 Saved Places.
> 
> **There are few places still for which the Google API does not return the coordinates**; some with a `q` param, and some with a `cid` param. Here is a GeoJSON that includes one of each:
> 
> ```json
> {
>   "type": "FeatureCollection",
>   "features": [
>     {
>       "geometry": {
>         "coordinates": [
>           0,
>           0
>         ],
>         "type": "Point"
>       },
>       "properties": {
>         "date": "2019-01-09T04:43:31Z",
>         "google_maps_url": "http://maps.google.com/?q=Unnamed+Road,+2+Chome+Ueno,+Tait%C5%8D-ku,+T%C5%8Dky%C5%8D-to+110-0005,+Japan&ftid=0x60188c20f3688631:0xfa27732134def7d4",
>         "Comment": "No location information is available for this saved place"
>       },
>       "type": "Feature"
>     },
>     {
>       "geometry": {
>         "coordinates": [
>           0,
>           0
>         ],
>         "type": "Point"
>       },
>       "properties": {
>         "date": "2018-08-01T18:53:27Z",
>         "google_maps_url": "http://maps.google.com/?cid=15807036382859735710",
>         "Comment": "No location information is available for this saved place"
>       },
>       "type": "Feature"
>     }
>   ]
> }
> ```
> 
> In the case of the first feature (with `q` param), the API returns ZERO_RESULTS. Since there is an `ftid` param provided, I tried to fetch that using the `https://maps.googleapis.com/maps/api/place/details/json` endpoint, which returned "NOT_FOUND: The provided Place ID is no longer valid. Please refresh cached Place IDs as per https://developers.google.com/places/web-service/place-id#save-id"
> 
> In the case of the second (with `cid` param), the API also returns the same message to refresh cached Place IDs. So we know that the issue with both features is an outdated Place ID.
> 
> I tried to see if we could solve that programmatically by following the instructions on the linked URL, which suggests refreshing stored place IDs using the Places API. Unfortunately, this did not prove fruitful:
> 
>     * Places API (New): `https://places.googleapis.com/v1/places/15807036382859735710?fields=id&key=` returns "Not a valid Place ID: 15807036382859735710"`
> 
>     * Places API: `https://maps.googleapis.com/maps/api/place/details/json? place_id=15807036382859735710?fields=place_id&key=` returns "Invalid placeid parameter"
> 
> 
> (Same result using the `ftid` from the first feature.)
> 
> I suspect that these instructions to refresh cached Place IDs don't apply to `ftid` or `cid`, so I'm not sure then how to access a valid Place ID for these.
> 
> Unless anyone has a suggestion, maybe we can leave this as an issue for follow-on work?

Great to hear that it worked. I think we can just skip the invalid place IDs and the user can add them manually if necessary.

**Review by fabwu — APPROVED** [review]
@rtsisyk The PR looks good to me and is ready to merge

@rudokemper Thanks four adding support for GeoJSON

**Comment by fabwu** [comment]
@rtsisyk Any chance to get this merged?

**Pull request merged**